### DISCUSSION TITLE
The basic idea

### DISCUSSION THREAD

**Discussion by jpivarski** [jpivarski/tiled-uproot#1] [General]
From an email I wrote to @danielballan:

In addition to reading and writing Awkward Arrays, there's another problem that we've been trying to solve for years: we have large datasets of ROOT files that would not be easy to convert permanently into another format, partly because of storage space. Uproot can transform ROOT files into Awkward Arrays, so there's a function that can put them into the same format as everything else, but that function needs to be used on-demand/lazily/virtually.

Martin Durant introduced me to https://github.com/fsspec/kerchunk, which makes a database of seek positions where the arrays are located within files, so that they can be accessed without having to go through the file format's own metadata, which is most of the problem: databases can look up seek positions much faster than file headers can be parsed.

I had the same idea about this problem years earlier; here's a proof-of-concept that I used to try to get the SkyHook developers interested: https://github.com/diana-hep/uproot-skyhook. I used FlatBuffers, while kerchunk uses JSON. ROOT files contain an especially large number of small arrays, so we wouldn't be able to use JSON for their locations. In a typical 2.4 GB ROOT file, the metadata of seek positions adds up to 11 MB; it's about 0.5% of the size of the data itself. For TB or PB-sized collections of ROOT files, we'd need a GB or TB-sized database.

It has just occurred to me that Tiled could be that database. This sounds too specific to be a part of the main Tiled project (but you can tell me!). I was thinking that another Python package could subclass or wrap the Tiled client to serve data from ROOT files as though they were Awkward Arrays—i.e. same interface, but when the data source is ROOT, it would just look up the locations in the ROOT files, extract the data, and serve that instead. It wouldn't function without both the database and access to the ROOT files.

What do you think? I could try setting up a local Tiled database, populating it with ROOT metadata and defining the procedures that go from the requested entry range, through the metadata-lookup, to extracting data from the ROOT files. Afterward, I'd like to get the interface to resemble the interface for reading and writing Awkward Arrays, so that users don't have to make a big distinction between directly-stored Awkward Arrays and lazy/virtual Awkward Arrays from ROOT. I may need to ask you detailed questions about how to do that.

**Comment by danielballan** [comment]
I completely missed this email! Searching for it, I see that I did in fact receive it, but missed it. I'm glad that you reiterated it here.

The use case you outline in the first paragraph resonates very much with the motivating ideas behind Tiled. This is exactly the kind of thing we are after, separating how data happens to be _stored_ from how it is accessed or presented:

>  serve data from ROOT files as though they were Awkward Arrays—i.e. same interface, but when the data source is ROOT, it would just look up the locations in the ROOT files, extract the data, and serve that instead.

Based on my understanding at this point, I think you could take the Tiled client and server as is, but devise a custom Tiled Adapter. Tiled Adapters (see `tiled.adapters`) are server-side entities. They abstract the specifics of reading from a given format or data source into the methods called by the Tiled server. Your Adapter could rely on a specialized database of offsets that suits your requirements. Tiled makes this plugable via a config file: "If you encounter a ROOT file, based on its file extension or file signature or some other indication, use this Adapter from this third-party module." It should be possible to get that working without specific accommodations in Tiled itself. Once it works, we can consider whether it makes sense to move this upstream, perhaps as a generalized capability that might be useful beyond ROOT---fast kerchunk-like lookup into HDF5 files?

One more thought: I remember a conversation we had at SciPy that touched on the possibility of placing the `structure` (form and length) in a separate table so that a given unique structure could be stored just once and then referenced, via foreign key, by the N datasets that happen to share that structure. I think you said that repeat structures were fairly common; that's probably true for our use cases as well. I won't rush into that optimization, but I think it is solidly in scope for Tiled.

**Reply by jpivarski** [reply]
Awesome! Tiled Adapters are what I'm looking for. In [the docs](https://blueskyproject.io/tiled/reference/service.html#adapters), it looks like Adapters are built into the main Tiled package, rather than being a plug-in system. Am I reading that right?

Tiled uses pip options for installation; `uproot` and `tiled-uproot` could become optional dependencies with a minimal Adapter wrapper in the Tiled codebase, or this repo could be an experiment that later gets integrated into Tiled.

> One more thought: I remember a conversation we had at SciPy that touched on the possibility of placing the `structure` (form and length) in a separate table so that a given unique structure could be stored just once and then referenced, via foreign key, by the N datasets that happen to share that structure.

That may be necessary for Awkward Arrays at some point in the future. For ROOT files, I'm already doing that: it's common for a suite of ROOT files to have almost the same set of columns, so I collect an "era" for each combination. Similarly, file names are collected as a single prefix, usually a common directory path, to avoid repeating the same bytes for each file (some manual compression, using the fact that we know a set of files are all in the same directory).

**Reply by danielballan** [reply]
We ship some adapters with tiled for "batteries included" reasons, but third-party ones can be registered and put on the same footing.

This will index a directory at server startup using a custom adapter:

```
tiled serve directory
  --verbose \
  --ext '.root=application/x-root' \
  --adapter 'application/x-root=custom_module:CustomAdapter' \
  data/
```

See also `--mimetype-hook` for more advanced mimetype detection. The above does the indexing at server startup in a temporary (SQLite) file. This is convenient for dev. For production deployments that are persistent and/or possibly horizontally scaled, you want to stand up a server with an empty persistent catalog and do the file-indexing as a distinct step.

There are docs for this approach. The docs on this could be improved, in terms of flow, because this is fairly new, but I think they are correct:
* https://blueskyproject.io/tiled/how-to/read-custom-formats.html
* https://blueskyproject.io/tiled/explanations/specialized-formats.html

I'm happy to clarify anything here, as well, of course.

**Reply by danielballan** [reply]
I spun on the conversation about storing `structure` separately: https://github.com/bluesky/tiled/issues/576

**Comment by jpivarski** [comment]
As an example of what I'm looking at, this is a typical dataset:

  * [/BTag/Run2011A-12Oct2013-v1/AOD](https://opendata.cern.ch/record/15)
  * [/BTag/Run2011B-12Oct2013-v1/AOD](https://opendata.cern.ch/record/270)

(`BTag` is the pre-filtered type of events, `2011A`-`2011B` is when the data were taken, and `12Oct2013` is when this sample was processed). Together, they're 2.3 TB, [674 files](https://opendata.cern.ch/record/15/files/CMS_Run2011A_BTag_AOD_12Oct2013-v1_20000_file_index.txt), and each file has about 1000‒2000 columns (depending on type), and each column is represented by 300‒600 contiguous chunks throughout the file. I'm trying to make a database of chunk seek locations and sizes so that I can get all chunks for a given set of columns in a range of events across all files in the dataset without having to find them by interpreting file headers with Python code.[^1]

On top of that, a typical data analysis needs about a dozen of these datasets, representing different trigger lines that would have captured the events of interest, as well as simulations of signal events and different types of backgrounds. The public [CERN data portal](https://opendata.cern.ch/search?page=1&size=20&experiment=CMS&file_type=aod&file_type=root&subtype=Collision&type=Dataset&collision_type=pp&sort=title) is serving about 3 PB of data from the LHC experiments, and unreleased datasets are bigger. I don't think we can index it all, but we may be able to index everything accessible at a particular site, which might be about a PB.

The metadata has a complex structure: a set of files like "`/BTag/Run2011A-12Oct2013-v1/AOD`" might have two or three eras (mentioned above), in which a column is added halfway through or a data type might change in a subtle way. (Each ROOT file describes its own data types.) Files are chunks, and there are chunks within the files, so it's a two-level jagged array of seek positions and lengths. Since the metadata is large, I'll need Awkward Arrays, rather than JSON, to represent it.

At present, I've settled on a data structure for the Awkward Array. It can be different for each ROOT dataset, since I want to make the column metadata separately addressable, and columns are different in each dataset, but it's homogeneous for a single dataset. (If a column is present in some files and absent in others, I use Awkward's option-type for the missing seek positions.)

Given the size of the metadata, it will be important to be able to fetch only parts of it. Tiled's slice-on-server feature is very helpful here! Fixing bluesky/tiled#574 will enable that, and it sounds like you already have that in scope.

I'll also need to be careful when talking about "metadata," since it can have several meanings here. The collision events in the ROOT files are data, the seek locations for extracting them as arrays are the metadata I'm using this post. Those metadata are Awkward Arrays, which have a Form and length, which are meta-metadata. That goes into Tiled, which has a catalog (SQLite in my configuration) alongside the files in a directory. I suppose that catalog is meta-meta-metadata?

[^1]: It's not just Python that makes finding the seek locations slow: the format is such that four interactions between client and server are needed to find them. (1) client asks for the first bytes in the file, the file header, (2) client uses the response to find the root directory inside the ROOT file, (3) client uses data from the directory to find the TTree metadata in the ROOT file, (4) client uses the TTree metadata to find the seek locations of interest. A high-latency network can be a big problem.

**Reply by danielballan** [reply]
Ha, I think I followed that. Makes sense so far.

Can you see more, then, about what a given ROOT file looks like to the Python client? Is it represented as a single AwkwardArray, or are you exposing it as a container of multiple AwkwardArrays (metadata and data)?

**Reply by jpivarski** [reply]
A collection of ROOT files could be presented as a big (TB scale) Awkward Array. The minor differences in which record fields are available across the dataset can be represented as missing values, so the whole collection would have one Form (with a few thousand fields).

Generally, people equals be downloading only a subset of fields and a subset of collision events (columns and rows), so the interface could be made to look exactly like an Awkward Array that Tiled is slicing server-side.

The interface that I'd like to see is one in which users don't care whether the data is sitting in ROOT files, only indexed by Tiled, or the data is directly served by Tiled. The Adapter interface you describe sounds like it can hide this detail. We'd keep the source data in ROOT files because of their size, but want derived data the users make would be in Awkward Arrays.

To go even further, it would be great if a single Awkward Array can be served from a mix of direct buffers and indirectly through ROOT files. That's in principle possible and would allow derived quantities to include unmodified source data without copying, but that's a more complicated next step.

**Reply by danielballan** [reply]
> The interface that I'd like to see is one in which users don't care whether the data is sitting in ROOT files, only indexed by Tiled, or the data is directly served by Tiled. 

Great, we are on the same page. For use cases at NSLS-II, we'll want one Tiled server to have a mixture of externally-managed data (typically written directly by detectors in whatever format) and internally-managed data (analysis results uploaded into Tiled by users). Client side, these will be indistinguishable.

> To go even further...

This is not quite implemented, but we have laid track for it. In the lingo of the tiled catalog SQL tables, this is one `node` (one logical `awkward` entity) backed by multiple `data_sources`, which happen to be a mixture of internally- and externally-managed. Again, the client side only sees the logical view.
